packages/labs/src/tagInput/tagInput.tsx

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ * Licensed under the BSD-3 License as modified (the “License”); you may obtain a copy
+ * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+ * and https://github.com/palantir/blueprint/blob/master/PATENTS
+ */
+
+import * as classNames from "classnames";
+import * as PureRender from "pure-render-decorator";
+import * as React from "react";
+
+import { AbstractComponent, Classes, HTMLInputProps, IProps, ITagProps, Keys, Tag, Utils } from "@blueprintjs/core";
+
+export interface ITagInputProps extends IProps {
+    /** React props to pass to the `<input>` element */
+    inputProps?: HTMLInputProps;
+
+    /**
+     * Callback invoked when a new tag is added by the user pressing `enter` on the input.
+     * Receives the current value of the input field. New tags are expected to be appended to
+     * the list.
+     */
+    onAdd?: (value: string) => void;
+
+    /**
+     * Callback invoked when the user clicks the X button on a tag.
+     * Receives value and index of removed tag.
+     */
+    onRemove?: (value: string, index: number) => void;
+
+    /**
+     * React props to pass to each `Tag`. Provide an object to pass the same props to every tag,
+     * or a function to customize props per tag.
+     *
+     * If you define `onRemove` here then you will have to implement your own tag removal
+     * handling as the `TagInput` `onRemove` prop will never be invoked.
+     */
+    tagProps?: ITagProps | ((value: string, index: number) => ITagProps);
+
+    /** Controlled tag values. */
+    values: string[];
+}
+
+export interface ITagInputState {
+    inputValue?: string;
+    isInputFocused?: boolean;
+}
+
+@PureRender
+export class TagInput extends AbstractComponent<ITagInputProps, ITagInputState> {
+    public static displayName = "Blueprint.TagInput";
+
+    public static defaultProps: Partial<ITagInputProps> = {
+        inputProps: {},
+        tagProps: {},
+    };
+
+    public state: ITagInputState = { inputValue: "", isInputFocused: false };
+
+    private inputElement: HTMLInputElement;
+    private refHandlers = {
+        input: (ref: HTMLInputElement) => {
+            this.inputElement = ref;
+            const refHandler = this.props.inputProps.ref;
+            if (Utils.isFunction(refHandler)) {
```

Comment:
seems there's not much to be done to improve `safeInvoke` usage. if i make it accept non-functions then it loses all type safety :(. adding code comment instead.